tests/cpp/operator/test_causal_softmax.cu

```diff
@@ -0,0 +1,253 @@
+/*************************************************************************
+ * Copyright (c) 2022-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See LICENSE for license information.
+ ************************************************************************/
+
+#include <cuda_bf16.h>
+#include <cuda_runtime.h>
+#include <gtest/gtest.h>
+
+#include <transformer_engine/softmax.h>
+#include "../test_common.h"
+
+using namespace transformer_engine;
+
+namespace {
+
+template <typename Type>
+void compute_single_head_fwd(
+  Type *softmax_out,
+  const Type *data_in,
+  const float scaling_factor,
+  const int rows,
+  const int cols)
+{
+  using compute_t = float;
+
+  for (int i = 0; i < rows; ++i) {
+    size_t offset = i * cols;
+
+    const int masked_elements = i + cols - rows + 1;
+    compute_t max_value = static_cast<compute_t>(-10'000.f);
+    for (int j = 0; j < masked_elements; ++j) {
+      compute_t tmp = scaling_factor * static_cast<compute_t>(data_in[offset + j]);
+      softmax_out[offset + j] = static_cast<Type>(tmp);
```

Comment:
We are incurring some numerical error by casting the intermediate state from `compute_t` to `Type`. It's not an enormous deal since softmax has good numerical stability, but allocating a temporary `compute_t` buffer would be better.

Reply:
The largest test case has sequence length 1024, which doesn't seem that long to me. Do you know which values are outside of the tolerances?

Reply:
I added more test cases which cover key sequence lengths up to 16384 (2^14). Two tests with bfloat16 fail for sequence lengths 256 and 1024:
```
FAILED: (bfloat16, 8, 6, 128, 256, 1.3)
Mismatch at place [1, 2, 18, 88] (266840): 0.00098419189453125 vs 0.000972747802734375

FAILED: (bfloat16, 1, 2, 819, 1024, 2.1)
Mismatch at place [0, 0, 0, 68] (68): 0.00140380859375 vs 0.0013885498046875
```

Reply:
The testing suite uses absolute tolerance `atol=1e-5` for bf16, which is too strict and is beyond the bf16 precision (`2^-7=7.8e-3`), relaxing to `atol=1e-3`